### ISSUE TITLE
Rack-aware load balancing

### ISSUE THREAD

**Issue by nyh** [scylladb/alternator-load-balancing#31]
Currently, all our Alternator load-balancing implementations in this repository ignore **rack** (a.k.a Amazon availability zone, AZ) information: We use the "/localnodes" API to get a list of all live Scylla servers in this data center (a.k.a Amazon region), and send the request to one of it. 

But when the Scylla DC has multiple _racks_ on different Amazon AZs,  cross-AZ traffic costs money. It is cheaper for the client running on a specific AZ to send the request to a random node on **the same** AZ - and not to nodes on other AZs. This issue requests that the load balancers do this: Prefer to send requests to a node on the client's rack, not a node on other racks.

See https://github.com/scylladb/scylladb/issues/12147 on a modification to "/localnodes" that can help us get the list of nodes in the current AZ.

Beyond server-size modifications the following two points will also need to be considered:

1. If the client itself is not _balanced_ across the different AZs (e.g., is only running on a single AZ), using current-AZ-only load balancing would be inefficient as it would lead the Scylla nodes on this AZ to be more loaded than others. If we can't recognize this situation automatically, we should at least make rack-aware load balancing _optional_.
2. If the cluster has just 3 nodes across 3 AZs, then each AZ has just a single node. If this node is temporarily down, listing the nodes on this AZ will return no node. If the current-AZ node list is empty, we should fall back to getting all the nodes in the whole DC, and use that list.

**Comment by nyh** [comment]
Oops., wrong repository :-( Closing.

**Comment by nyh** [comment]
@dkropachev this is true - we have https://github.com/scylladb/alternator-load-balancing/issues/11 for token-aware load balancer, but as I noted there, there is a difficulty: it will mean we'll need to monkey-patch the AWS SDK at a different place than we do today, to let it see the full query - and parse it (unfortunately) - to decide where to route a write request (for reads, LWT is not relevant).

We also have https://github.com/scylladb/scylladb/issues/5703 on the Scylla side, which says that if the AWS SDK isn't token-aware (like it isn't today), we can rescue the contention problem by forward writes to the "right" node.

But you're right - if the load balancer is rack-aware (as this issue proposes), and different racks will send writes to different nodes, we will end up with **more** LWT contention. I don't know what to do about this - other than making rack-aware load balancing optional. Personally, I think the LWT whole-partition-contention problems need to be fixed (https://github.com/scylladb/scylladb/issues/16261) instead of trying to work around them in the load balancer.

**Comment by mykaul** [comment]
CC @kostja - thoughts?

**Comment by dkropachev** [comment]
if we have plans to enable LWT on tablets, then either we need to bring drivers to the alternator load balancers, or we need to expose routing info via API.

**Comment by mykaul** [comment]
@dkropachev - any progress on this? (regardless of tablets)

**Comment by dkropachev** [comment]
@mykaul, it is scheduled to the next sprint

**Comment by mykaul** [comment]
> [@mykaul](https://github.com/mykaul), it is scheduled to the next sprint

It'd be great if it can be prioritized and delivered sooner. It has a material impact.

**Comment by mykaul** [comment]
> > [@mykaul](https://github.com/mykaul), it is scheduled to the next sprint
> 
> It'd be great if it can be prioritized and delivered sooner. It has a material impact.

@dkropachev , @roydahan - what's the status of this?

**Comment by nyh** [comment]
> [@dkropachev](https://github.com/dkropachev) , [@roydahan](https://github.com/roydahan) - what's the status of this?

@mykaul do you know which of the languages and SDK versions that we support you want this feature to appear in **first**?

**Comment by mykaul** [comment]
> > [@dkropachev](https://github.com/dkropachev) , [@roydahan](https://github.com/roydahan) - what's the status of this?
> 
> [@mykaul](https://github.com/mykaul) do you know which of the languages and SDK versions that we support you want this feature to appear in **first**?

Java for sure. Not sure about SDK version.

**Comment by roydahan** [comment]
The first PRs are only for Java.

**Comment by dkropachev** [comment]
> > [@dkropachev](https://github.com/dkropachev) , [@roydahan](https://github.com/roydahan) - what's the status of this?
> 
> [@mykaul](https://github.com/mykaul) do you know which of the languages and SDK versions that we support you want this feature to appear in **first**?

Java, both SDK versions.

**Comment by nyh** [comment]
Implemented in Java, see pull request https://github.com/scylladb/alternator-load-balancing/issues/40.
Not yet implemented in other languages.

**Comment by dani-tweig** [comment]
Closing this issue as it was moved to Jira.
Please continue the thread in https://scylladb.atlassian.net/browse/DRIVER-26
